fix(term): ensure terminal attr index doesn't exceed `TERM_ATTRS_MAX`

The-Briel-Deal opened the pull request:
Problem: Currently terminal highlight attribute buffers are statically allocated
be the size of `TERM_ATTRS_MAX`. This unique case isn't respected in
some places in the ui_compositor. Due to this, when a terminal window
has lines longer them `TERM_ATTRS_MAX`, the compositor will go past the
end of the buffer causing a crash due to out of bounds access.

Solution: Add check to ensure we don't query terminal highlight attrs
past `TERM_ATTRS_MAX` in `win_line()`.

Fixes https://github.com/neovim/neovim/issues/30374

The-Briel-Deal commented on src/nvim/ui_compositor.c:
```diff
@@ -9,6 +9,7 @@
 #include <stdbool.h>
 #include <stdlib.h>
 #include <string.h>
+#include <sys/param.h>
```
Sorry about that! I just dropped it, that was from lsp auto-include.

justinmk commented on src/nvim/ui_compositor.c:
```diff
@@ -591,11 +591,15 @@ void ui_comp_raw_line(Integer grid, Integer row, Integer startcol, Integer endco
     compose_debug(row, row + 1, endcol, clearcol, dbghl_clear, true);
 #ifndef NDEBUG
-    for (int i = 0; i < endcol - startcol; i++) {
+    Integer cols = endcol - startcol;
+    if (max_attrs != 0) {
+      cols = MIN(cols, max_attrs - 1);
+    }
```
```suggestion
    Integer cols = max_attrs == 0 ? endcol - startcol : MIN(cols, max_attrs - 1);
```

justinmk commented on src/nvim/grid.h:
```diff
@@ -30,9 +30,12 @@ EXTERN char *linebuf_scratch INIT( = NULL);
+  SLF_RIGHTLEFT   = 1,
+  SLF_WRAP        = 2,
+  SLF_INC_VCOL    = 4,
+  SLF_TERM_ATTRS  = 8,  // On if this line is associated with a terminal buffer. If so then we need
+                        // to ensure we don't go past index `TERM_ATTRS_MAX` in the highlight
+                        // attributes.
```
```suggestion
  SLF_TERMATTRS  = 8,  // Indicates the line is in a terminal buffer, so we must not exceed
                       // index `TERM_ATTRS_MAX` in the highlight attributes.
```

The-Briel-Deal replied:
Ack, i'll apply this once I'm off work later today.

justinmk commented on src/nvim/ui_compositor.c:
```diff
@@ -591,11 +591,15 @@ void ui_comp_raw_line(Integer grid, Integer row, Integer startcol, Integer endco
-    for (int i = 0; i < endcol - startcol; i++) {
+    Integer cols = endcol - startcol;
+    if (max_attrs != 0) {
+      cols = MIN(cols, max_attrs - 1);
```
Is this the only place where `max_attrs` is used? This code is only for debugging (`#ifndef NDEBUG`).

And even if it wasn't, adding a parameter in many places, and setting a special `SLF_TERM_ATTRS` flag per-line, does not seem like the right fix.

justinmk replied:
> I believe (I would have to look into this deeper) that for other buffers, there will always be the same number of columns as highlight attrs for a line.
> 
> ...
>
> I couldn't figure out why the terminal is the one buffer type where we can have less highlight attrs the columns.

@bfredl may have insight to both of those.



> The final option which in my mind is ideal is to handle terminal highlight attrs the same way we handle highlight attrs in other buffers so that we don't have this special case

ideally yes, though I may be missing something.

The-Briel-Deal replied:
Let me look into just changing the terminal to mirror the behavior of highlight attrs on other buffers. @bfredl if you have any insight into why terminals have a cap on the number of highlight attrs while other buffers don't that would be super helpful.

Thanks for the feedback Justin! I had the same feeling about the amount of code I had to add for this small check. I'm going to make this CL draft for now until I can figure out a better solution.

The-Briel-Deal commented:
> > This will be set to `TERM_ATTRS_MAX` for terminals, and 0 if there is no limit.
> 
> There is always a limit, but it might be much bigger for the non-terminal case.
> 
> Wondering if there is a different way for this limit to be signaled to the UI, instead of passing it around everywhere (which is pretty redundant, and rather spread-out).

I totally agree, i'm just not familiar enough with the codebase to know how else I can pass that info to the UI. Maybe a buffer local var?

The-Briel-Deal committed:
fix(term): ensure terminal attr index doesn't exceed `TERM_ATTRS_MAX`

Problem: Currently terminal highlight attribute buffers are statically allocated
be the size of `TERM_ATTRS_MAX`. This unique case isn't respected in
some places in the ui_compositor. Due to this, when a terminal window
has lines longer them `TERM_ATTRS_MAX`, the compositor will go past the
end of the buffer causing a crash due to out of bounds access.

Solution: Add check to ensure we don't query terminal highlight attrs
past `TERM_ATTRS_MAX` in `win_line()`.

Fixes #30374

zeertzjq commented on src/nvim/drawline.c:
```diff
@@ -2265,7 +2265,7 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, int col_rows, b
 
         if (wp->w_buffer->terminal) {
-          wlv.char_attr = hl_combine_attr(term_attrs[wlv.vcol], wlv.char_attr);
+          wlv.char_attr = hl_combine_attr(TERM_ATTRS_MAX > wlv.vcol ? term_attrs[wlv.vcol] : 0, wlv.char_attr);
```
This is a bit hard to read
```suggestion
          wlv.char_attr = hl_combine_attr(wlv.vcol < TERM_ATTRS_MAX ? term_attrs[wlv.vcol] : 0, wlv.char_attr);
```

The-Briel-Deal replied:
Thats fair, i'll fix that

justinmk commented on test/functional/terminal/highlight_spec.lua:
```diff
@@ -408,3 +409,23 @@ describe(':terminal', function()
   end)
 end)
+
+describe(':terminal max highlight attrs', function()
+  it('does not cause crash on zoomout with large horizontal output #30374', function()
```
can this just be an it() in one of the existing describe() blocks, why do we need a new describe()

The-Briel-Deal replied:
We don't! I'll drop that!

justinmk reviewed (approved):
nice, thanks!

justinmk commented on test/functional/terminal/highlight_spec.lua:
```diff
@@ -408,3 +409,21 @@ describe(':terminal', function()
   end)
 end)
+
+it(':terminal does not cause crash on zoomout with large horizontal output #30374', function()
```
this should be in one of the existing describe() blocks : https://github.com/neovim/neovim/blob/4ee2e365a583e26468e7ea820f5416a980515504/test/functional/terminal/highlight_spec.lua#L392

The-Briel-Deal replied:
Done! Ty

justinmk commented on test/functional/terminal/highlight_spec.lua:
```diff
@@ -408,3 +409,21 @@ describe(':terminal', function()
   end)
 end)
+
+it(':terminal does not cause crash on zoomout with large horizontal output #30374', function()
```
```suggestion
it('zoomout with large horizontal output #30374', function()
```

The-Briel-Deal commented:
Thank you @justinmk! I misread your comment about moving to an existing describe block. All comments should be addressed now!
